dra-evolution/pkg/api/claim_types.go

```diff
@@ -1,379 +1,670 @@
-	// chosen devices, without having to explicitly specify the value of
-	// the constraint.  For example, this allows constraints like "all
-	// devices must be the same model", without having to specify the exact
-	// model. We may be able to use this for some basic topology
-	// constraints too, by representing the topology as device attributes.
+	// Allocation is set once the claim has been allocated successfully.
+	// +optional
+	Allocation *AllocationResult `json:"allocation,omitempty" protobuf:"bytes,2,opt,name=allocation"`
+
+	// ReservedFor indicates which entities are currently allowed to use
+	// the claim. A Pod which references a ResourceClaim which is not
+	// reserved for that Pod will not be started.
 	//
-	// Currently, these are just strings. However, we could make them
-	// structs, and include required vs preferred matches. Required matches
-	// would fail if not met, where as preferred would lower the score if
-	// not met. We could even allow low/medium/high priority and adjust the
-	// score differently for each.
+	// There can be at most 32 such reservations. This may get increased in
+	// the future, but not reduced.
 	//
+	// +listType=map
+	// +listMapKey=uid
+	// +patchStrategy=merge
+	// +patchMergeKey=uid
 	// +optional
-	// +listType=atomic
-	MatchAttributes []string `json:"matchAttributes,omitempty"`
+	ReservedFor []ResourceClaimConsumerReference `json:"reservedFor,omitempty" protobuf:"bytes,3,opt,name=reservedFor" patchStrategy:"merge" patchMergeKey:"uid"`
 
-	// Configs contains references to arbitrary vendor device configuration
-	// objects that will be attached to the device allocation.
+	// DeallocationRequested indicates that a ResourceClaim is to be
+	// deallocated.
+	//
+	// The driver then must deallocate this claim and reset the field
+	// together with clearing the Allocation field.
+	//
+	// While DeallocationRequested is set, no new consumers may be added to
+	// ReservedFor.
+	//
+	// This is an alpha field and requires enabling the DRAControlPlaneController
+	// feature gate.
+	//
 	// +optional
-	// +listType=atomic
-	Configs []DeviceConfigReference `json:"configs,omitempty"`
+	DeallocationRequested bool `json:"deallocationRequested,omitempty" protobuf:"varint,4,opt,name=deallocationRequested"`
 }
 
-// DeviceClaimStatus contains the results of the claim allocation.
-type DeviceClaimStatus struct {
-	// ClassConfigs contains the entire set of dereferenced vendor
-	// configurations from the DeviceClass, as of the time of allocation.
-	// +optional
+// AllocationResult contains attributes of an allocated resource.
+type AllocationResult struct {
+	// DriverData contains the state associated with an allocation that
+	// should be maintained throughout the lifetime of a claim. Each
+	// entry contains data that should be passed to a specific kubelet
+	// plugin once the claim lands on a node.
+	//
+	// Setting this field is optional. It has a maximum size of 32 entries.
```

Comment:
Where does 32 come from? Is it arbitrary?

Reply:
Yes, arbitrary. Lists should not have unlimited size, so we had to pull some number out of a hat when doing the initial DRA API.

The same applies also to all other lists that we are adding now. We just haven't gotten to that level of detail yet :cry: